### PR TITLE
Fixes #8587 - Don't allow build for unmanaged hosts

### PR DESCRIPTION
**Pull request by adamruzicka**

### CODE REVIEW

File: app/models/host/base.rb [hunk 1]
```diff
@@ -469,6 +470,12 @@ def uniq_interfaces_identifiers
       success
     end
 
+    def build_managed_only
+      (!self.build? || self.managed?).tap do |valid|
```
Comment (domcleal) [inline app/models/host/base.rb]: Why use `tap` and `unless valid` (line below) here?  This looks like it should be a simple conditional:

```
if !managed? && build?
  errors.add(...)
end
```

File: test/unit/host_test.rb [hunk 2]
```diff
@@ -1032,6 +1032,21 @@ def teardown
       assert h.errors[:root_pass].include?("should be 8 characters or more")
     end
 
+    test "should allow build mode for managed hosts" do
+      h = FactoryGirl.create(:host, :managed)
```
Comment (domcleal) [inline test/unit/host_test.rb]: use .build here, no need to save the host

File: test/unit/host_test.rb [hunk 3]
```diff
@@ -1032,6 +1032,21 @@ def teardown
...
+      assert h.valid?
+      h.build = true
+      assert h.save!
```
Comment (domcleal) [inline test/unit/host_test.rb]: Only assert validity to avoid saving the host and hitting the DB.

**Comment by domcleal** [comment]
There are some failures in other tests, please address them too.

File: test/unit/orchestration/tftp_test.rb [hunk 4]
```diff
@@ -30,10 +30,9 @@ def test_host_should_not_have_tftp
 
   def test_generate_pxe_template_for_build
     if unattended?
-      h = FactoryGirl.create(:host, :build => true,
-                             :operatingsystem => operatingsystems(:redhat),
```
Comment (domcleal) [inline test/unit/orchestration/tftp_test.rb:34]: If you're removing these to work around validation when creating a managed host, I think you can simply switch this FG call to .build - which both keeps it neater, and significantly reduces the number of records created.

File: app/models/host/base.rb [hunk 5]
```diff
@@ -469,6 +470,12 @@ def uniq_interfaces_identifiers
       success
     end
 
+    def build_managed_only
+      if !managed? && build?
+        errors.add(:build, _('cannot be enabled for unmanaged host'))
```
Comment (domcleal) [inline app/models/host/base.rb]: `[..] for an unmanaged host`

File: test/unit/orchestration/tftp_test.rb (hunk 4 above)
Comment (adamruzicka) [inline test/unit/orchestration/tftp_test.rb:34]: That was exactly the reason

**Comment by domcleal** [comment]
Merged as dac1677, thanks @adamruzicka.

**Pull request closed without merging**